tpcc/check.go

```diff
@@ -3,129 +3,274 @@ package main
-	var sumOOLCnt, countNewOrderLine int
-	if err := forEachWarehouseAndDistrict(func(wID, dID int) error {
-		if err := db.QueryRow(`SELECT sum(o_ol_cnt) FROM "order" WHERE o_w_id=$1 AND o_d_id=$2`,
-			wID, dID,
-		).Scan(&sumOOLCnt); err != nil {
-			return err
+	var i int
+	var val int
+	for ; rows.Next(); i++ {
+		rows.Scan(&val)
+		if val != -1 {
+			return errors.Errorf("MAX(no_o_id) - MIN(no_o_id) - COUNT(rows) must be -1, got %d", val)
 		}
-		if err := db.QueryRow(`SELECT count(*) FROM order_line WHERE ol_w_id=$1 AND ol_d_id=$2`,
-			wID, dID,
-		).Scan(&countNewOrderLine); err != nil {
-			return err
-		}
-		if countNewOrderLine != sumOOLCnt {
-			fmt.Printf("check failed: count(order_line)=%d != sum(o_ol_cnt)=%d for w_id=%d d_id=%d\n",
-				countNewOrderLine, sumOOLCnt, wID, dID)
-		}
-		return nil
-	}); err != nil {
+	}
+
+	if err := rows.Close(); err != nil {
 		return err
 	}
 
-	// 3.3.2.8
-	// W_YTD = sum(H_AMOUNT) for each warehouse
-	var sumHAmount float64
-	for i := 0; i < *warehouses; i++ {
-		if err := db.QueryRow("SELECT w_ytd FROM warehouse WHERE w_id=$1", i).Scan(&wYTD); err != nil {
-			return err
+	if i == 0 {
+		return errors.Errorf("zero rows")
+	}
+
+	fmt.Printf("3.3.2.3: %d rows, passed in %f seconds.\n", i, time.Since(start).Seconds())
+	return nil
+}
+
+func check3324(db *sql.DB) error {
+	start := time.Now()
+
+	leftRows, err := db.Query("SELECT SUM(o_ol_cnt) FROM tpcc.order GROUP BY o_w_id, o_d_id ORDER BY o_w_id, o_d_id")
+	if err != nil {
+		return err
+	}
+	rightRows, err := db.Query("SELECT count(*) from tpcc.order_line GROUP BY ol_w_id, ol_d_id ORDER BY ol_w_id, ol_d_id")
+	if err != nil {
+		return err
+	}
+	var i int
+	var left, right int64
+	for ; leftRows.Next() && rightRows.Next(); i++ {
+		leftRows.Scan(&left)
+		rightRows.Scan(&right)
+		if left != right {
+			return errors.Errorf("order.SUM(o_ol_cnt): %d != order_line.count(*): %d", left, right)
 		}
-		if err := db.QueryRow("SELECT SUM(h_amount) FROM history WHERE h_w_id=$1", i).Scan(&sumHAmount); err != nil {
-			return err
+	}
+	if i == 0 {
+		return errors.Errorf("0 rows returned")
+	}
+	if leftRows.Next() || rightRows.Next() {
+		return errors.Errorf("length of order.SUM(o_ol_cnt) != order_line.count(*)")
+	}
+
+	if err := leftRows.Close(); err != nil {
+		return err
+	}
+	if err := rightRows.Close(); err != nil {
+		return err
+	}
+
+	fmt.Printf("3.3.2.4: %d rows, passed in %f seconds.\n", i, time.Since(start).Seconds())
+	return nil
+}
+
+func check3325(db *sql.DB) error {
+	start := time.Now()
+
+	// We want the symmetric difference between the sets:
+	// (SELECT no_w_id, no_d_id, no_o_id FROM tpcc.new_order)
+	// (SELECT o_w_id, o_d_id, o_id FROM tpcc.order@primary WHERE o_carrier_id IS NULL)
+	// We achieve this by two EXCEPT ALL queries.
+
+	firstQuery, err := db.Query("(SELECT no_w_id, no_d_id, no_o_id FROM tpcc.new_order) EXCEPT ALL (SELECT o_w_id, o_d_id, o_id FROM tpcc.order@primary WHERE o_carrier_id IS NULL)")
```

Comment:
Nit: you can add newlines to these queries to make them more readable. For example:

```
`(SELECT no_w_id, no_d_id, no_o_id FROM tpcc.new_order) 
EXCEPT ALL 
(SELECT o_w_id, o_d_id, o_id FROM tpcc.order@primary WHERE o_carrier_id IS NULL)`
```

(Note that this is a back-quoted string)